tools/autograd/gen_variable_type.py

```diff
@@ -82,6 +82,12 @@
     '_coalesced_',
 }
 
+# Some Variables are not allowed to have their size or storage changed, if they
+# share the same storage with another Variable (e.g. via `.data` or `detach()`)
+CHECK_ALLOW_SIZE_OR_STORAGE_CHANGE = {
+    'set_',
```

Comment:
I don't think we should hard code these -- can we just make sure that all functions that modify the shape of a TensorImpl go through a single function and check there?